File: src/mt/edu/um/cf2/jgribx/Logger.java
/*
 * ============================================================================
 * JGribX
 * ============================================================================
 * Written by Andrew Spiteri <andrew.spiteri@um.edu.mt>
 * Adapted from JGRIB: http://jgrib.sourceforge.net/
 * 
 * Licensed under MIT: https://github.com/spidru/JGribX/blob/master/LICENSE
 * ============================================================================
 */
package mt.edu.um.cf2.jgribx;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

public class Logger
{
    public enum LoggingMode {OFF,CONSOLE,LOCAL,REMOTE};
    private static LoggingMode mode;
    
    private static PrintWriter pw;
    
    /** Designates severe error events that typically lead to application abort. */
    public static final int FATAL   = 0;
    /** Designates error events that might still allow the application to continue running. */
    public static final int ERROR   = 1;
    /** Designates potentially problematic events. */
    public static final int WARNING = 2;
    /** Designates informational messages that highlight the progress of the application. */
    public static final int INFO    = 3;
    /** Designates fine-grained informational events that are mostly useful for debugging. */
    public static final int DEBUG   = 4;
    /** Designates finer-grained informational events that are useful for tracing a problem. */
    public static final int TRACE   = 5;
    
    private static int level;
    
    public static void setLoggingMode(LoggingMode mode)
    {
        Logger.mode = mode;
    }
    
    public static void setLevel(int lvl)
    {
         level = lvl;
    }
    
    public static void print(String msg, int lvl)
    {
        if (lvl > level) return;
        switch (lvl)
        {
            case FATAL:
                msg = "FATAL: " + msg;
                break;
            case ERROR:
                msg = "ERROR: " + msg;
                break;
            case WARNING:
                msg = "WARNING: " + msg;
                break;
            case INFO:
                msg = "INFO: " + msg;
                break;
            case DEBUG:
                msg = "DEBUG: " + msg;
                break;
            case TRACE:
                msg = "TRACE: " + msg;
                break;
            default:
                break;
        }
        switch (mode)
        {
            case OFF:
                break;
            case CONSOLE:
                System.out.print(msg);
                break;
            case LOCAL:
                if (pw == null)
                    pw = openLogFile(new File("log.txt"));
                pw.print(msg);
                break;
            case REMOTE:
                throw new UnsupportedOperationException("The current version of Logger does not yet support remote logging");
        }
    }
    
    public static void println(String msg, int lvl)
    {
        print(msg+"\r\n",lvl);
    }
    
    public static void println(int lvl)
    {
        print("\r\n",lvl);
    }
    
    public static void flush()
    {
        if (pw != null)
            pw.flush();
    }
    
    private static PrintWriter openLogFile(File file)
    {
        PrintWriter pw = null;
        if (!file.exists())
        {
            try
            {
                file.createNewFile();
                System.out.println("New log file created");
            }
            catch (IOException ioe)
            {
                System.out.println("Cannot create log file");   // FIXME always writing to stdout
            }
        }
        try
        {
            pw = new PrintWriter(new BufferedWriter(new FileWriter(file,true)));
        }
        catch (IOException ioe)
        {
            // TODO Add exception handler
        }
        return pw;
    }
}

File: src/main/java/mt/edu/um/cf2/jgribx/Logger.kt
/*
 * ============================================================================
 * JGribX
 * ============================================================================
 * Written by Andrew Spiteri <andrew.spiteri@um.edu.mt>
 * Adapted from JGRIB: http://jgrib.sourceforge.net/
 * 
 * Licensed under MIT: https://github.com/spidru/JGribX/blob/master/LICENSE
 * ============================================================================
 */
package mt.edu.um.cf2.jgribx

import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.io.PrintWriter

object Logger {
	enum class LoggingMode {
		OFF, CONSOLE, LOCAL, REMOTE
	}

	/** Designates severe error events that typically lead to application abort.  */
	const val FATAL = 0

	/** Designates error events that might still allow the application to continue running.  */
	const val ERROR = 1

	/** Designates potentially problematic events.  */
	const val WARNING = 2

	/** Designates informational messages that highlight the progress of the application.  */
	const val INFO = 3

	/** Designates fine-grained informational events that are mostly useful for debugging.  */
	const val DEBUG = 4

	/** Designates finer-grained informational events that are useful for tracing a problem.  */
	const val TRACE = 5

	private var pw: PrintWriter? = null
	var mode: LoggingMode? = null
	var level = 0

	fun print(message: String, lvl: Int) {
		if (lvl > level) return
		val formattedMessage = when (lvl) {
			FATAL -> "FATAL: ${message}"
			ERROR -> "ERROR: ${message}"
			WARNING -> "WARNING: ${message}"
			INFO -> "INFO: ${message}"
			DEBUG -> "DEBUG: ${message}"
			TRACE -> "TRACE: ${message}"
			else -> message
		}
		when (mode) {
			LoggingMode.OFF -> {
			}
			LoggingMode.CONSOLE -> print(formattedMessage)
			LoggingMode.LOCAL -> {
				pw = (pw ?: openLogFile(File("log.txt")))
						?.also { it.print(formattedMessage) }
			}
			LoggingMode.REMOTE -> throw UnsupportedOperationException("The current version of Logger does not yet support remote logging")
		}
	}

	fun debug(message: String?, throwable: Throwable? = null) = println(message, throwable, DEBUG)

	fun info(message: String?, throwable: Throwable? = null) = println(message, throwable, INFO)

	fun warning(message: String?, throwable: Throwable? = null) = println(message, throwable, WARNING)

	fun error(message: String?, throwable: Throwable? = null) = println(message, throwable, ERROR)

	private fun println(message: String?, throwable: Throwable?, lvl: Int) {
		print("${message ?: ""}\n", lvl)
		throwable?.printStackTrace(if (lvl < 2) System.err else System.out)
	}

	fun flush() {
		pw?.flush()
	}

	private fun openLogFile(file: File): PrintWriter? {
		if (!file.exists()) try {
			file.createNewFile()
			println("New log file created")
		} catch (ioe: IOException) {
			println("Cannot create log file") // FIXME always writing to stdout
		}
		return try {
			PrintWriter(BufferedWriter(FileWriter(file, true)))
		} catch (ioe: IOException) {
			// TODO Add exception handler
			null
		}
	}
}